src/cargo/ops/cargo_new.rs

```diff
@@ -67,22 +64,144 @@ pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
+            
+        }
+        opts2.sourcefile_relative_path = Some(format!("{}.rs", name));
+        found_source_files += 1;
+    }
+    
+    if paths::file_already_exists(&path.join(format!("src/{}.rs", name))) {
+        if opts2.bin {
+            // OK
+        } else {
+            opts2.bin = try!(autodetect_bin_file(&path.join(format!("src/{}.rs", name))));
+            
+        }
+        opts2.sourcefile_relative_path = Some(format!("src/{}.rs", name));
+        found_source_files += 1;
+    }
+    
+    if (!opts2.bin) && paths::file_already_exists(&path.join("src/lib.rs")) {
+        found_source_files += 1;
+        opts2.sourcefile_relative_path = Some(String::from("src/lib.rs"));
+    }
+    if (!opts2.bin) && paths::file_already_exists(&path.join("lib.rs")) {
+        found_source_files += 1;
+        opts2.sourcefile_relative_path = Some(String::from("lib.rs"));
+    }
+    
+    if found_source_files > 1 {
+        return Err(human(r"There are multiple eligible source files for `cargo init`.
+I don't know which Cargo.toml template to use."));
+    }
+    Ok(())
+}
+
+pub fn new(opts: NewOptions, config: &Config) -> CargoResult<()> {
+    let path = config.cwd().join(opts.path);
+    if fs::metadata(&path).is_ok() {
+        return Err(human(format!("Destination `{}` already exists",
+                                 path.display())))
+    }
+    
+    let name = try!(get_name(&path, &opts, config));
+    try!(check_name(name));
+    
     mk(config, &path, name, &opts).chain_error(|| {
         human(format!("Failed to create project `{}` at `{}`",
                       name, path.display()))
     })
 }
 
+pub fn init(opts: NewOptions, config: &Config) -> CargoResult<()> {
+    assert_eq!(opts.path, ".");
+    let path = config.cwd().join(opts.path);
+    
+    let cargotoml_path = path.join("Cargo.toml");
+    if fs::metadata(&cargotoml_path).is_ok() {
+        return Err(human(format!("Destination `{}` already exists",
+                                 cargotoml_path.display())))
+    }
+    
+    let name = try!(get_name(&path, &opts, config));
+    try!(check_name(name));
+    
+    let mut opts2 = opts.clone();
+    
+    if opts2.sourcefile_relative_path == None {
+        try!(detect_source_path_and_type(&path, name, &mut opts2));
+    }
+    
+    if opts2.version_control == None {
+        let mut num_detected_vsces = 0;
+        
+        if fs::metadata(&path.join(".git")).is_ok() {
+            opts2.version_control = Some(VersionControl::Git);
+            num_detected_vsces += 1;
+        }
+        
+        if fs::metadata(&path.join(".hg")).is_ok() {
+            opts2.version_control = Some(VersionControl::Hg);
+            num_detected_vsces += 1;
+        }
+        
+        // if none exists, maybe create git, like in `cargo new`
+        
+        if num_detected_vsces > 1 {
+            return Err(human("Both .git and .hg exist. I don't know what to choose."));
```

Comment:
Hm this error seems kinda odd, perhaps cargo init could just not do anything VCS-related if no VCS was found? If at least one was found it could just skip that step.